crossbeam-channel/src/flavors/array.rs

```diff
@@ -312,6 +313,88 @@ impl<T> Channel<T> {
         }
     }
 
+    fn start_recv_many<E: Extend<T>>(&self, buf: &mut E, limit: usize) -> Result<usize, TryRecvError> {
+        let mut backoff = Backoff::new();
+
+        loop {
+            // Load the head/tail and deconstruct them.
+            let head = self.head.load(Ordering::SeqCst);
+            let tail = self.tail.load(Ordering::SeqCst);
+            let head_index = head & (self.one_lap - 1);
+            let head_lap = head & !(self.one_lap - 1);
+            let tail_index = tail & (self.one_lap - 1);
+
+            // Inspect the corresponding head slot.
+            let head_slot = unsafe { &*self.buffer.add(head_index) };
+            let head_stamp = head_slot.stamp.load(Ordering::Acquire);
+
+            // If the head and the head stamp match, we may attempt to pop.
+            if head == head_stamp {
+                // See how many messages we have available to read.
+                let msgs_avail = if head_index < tail_index {
+                    tail_index - head_index
+                } else {
+                    self.cap - head_index + tail_index
+                };
+
+                // But respect the configured limit.
+                let msgs_to_count = cmp::min(limit, msgs_avail);
+
+                // Now, to reserve this whole range, we have to update head to be after our range end.
+                let new_head = if head_index + msgs_to_count < self.cap {
+                    head + msgs_to_count
+                } else {
+                    // We calculate the difference after wrapping around our capacity, and add that
+                    // to the index to land in the right slot next time.
+                    let diff = head_index + msgs_to_count - self.cap;
+                    head_lap.wrapping_add(self.one_lap.wrapping_mul(2) + diff)
+                };
+
+                let new_head_index = new_head & (self.one_lap - 1);
+
+                // Try moving the head.
+                if self
+                    .head
+                    .compare_exchange_weak(head, new_head, Ordering::SeqCst, Ordering::Relaxed)
+                    .is_ok()
+                {
+                    // We've claimed the full range of messages.  Read each one and store it.
+                    let mut index = head_index;
+                    while index != new_head_index {
+                        let slot = unsafe { &*self.buffer.add(index) };
+                        let msg = unsafe { slot.msg.get().read() };
```

Comment:
I believe the problem lies here.

Recall that:
* If `head == head_stamp`, that means a value has been written into the slot.
* If `head != head_stamp`, that means a value is not written into the slot. If the queue is not empty, a value will be written very soon, so we just need to spin-wait until `head == head_stamp`.

Before this while loop, we've only checked that the first slot is safe for reading. But for each of the following slots, we have to spin-wait until the stamp indicates that the value has been written into the slot.